crates/pallet-domains/src/bundle_storage_fund.rs

```diff
@@ -0,0 +1,107 @@
+use frame_support::traits::fungible::Mutate;
+use frame_support::traits::tokens::{Fortitude, Precision};
+use frame_support::traits::Get;
+use frame_support::PalletError;
+use scale_info::TypeInfo;
+use sp_domains::OperatorId;
+use sp_runtime::traits::{AccountIdConversion, CheckedSub, Zero};
+use sp_runtime::Perbill;
+use sp_std::collections::btree_map::BTreeMap;
+use subspace_runtime_primitives::StorageFeeInterface;
+
+/// Bundle storage fund specific errors
+#[derive(TypeInfo, Encode, Decode, PalletError, Debug, PartialEq)]
+pub enum Error {
+    FailedToDeriveStorageFundAccount,
+    BundleStorageFeePayment,
+    BalanceUnderflow,
+    MintBalance,
+}
+
+/// The type of system account being created.
+#[derive(Encode, Decode)]
+pub enum AccountType {
+    StorageFund,
+}
+
+/// Return the bundle storage fund account of the given operator.
+pub fn storage_fund_account<T: Config>(id: OperatorId) -> Result<T::AccountId, Error> {
+    T::PalletId::get()
+        .try_into_sub_account((AccountType::StorageFund, id))
+        .ok_or(Error::FailedToDeriveStorageFundAccount)
+}
+
+/// Charge the bundle storage fee from the operator's bundle storage fund
+pub fn charge_bundle_storage_fee<T: Config>(
+    operator_id: OperatorId,
+    bundle_size: u32,
+) -> Result<(), Error> {
+    if bundle_size.is_zero() {
+        return Ok(());
+    }
+
+    let storage_fund_acc = storage_fund_account::<T>(operator_id)?;
+    let storage_fee = T::StorageFeeInterface::transaction_byte_fee() * bundle_size.into();
+
+    T::Currency::burn_from(
+        &storage_fund_acc,
+        storage_fee,
+        Precision::Exact,
+        Fortitude::Polite,
+    )
+    .map_err(|_| Error::BundleStorageFeePayment)?;
+
+    // Note the storage fee, it will go to the consensus block author
+    T::StorageFeeInterface::note_storage_fees(storage_fee);
+
+    Ok(())
+}
+
+/// Refund the front paid storage fee of a particular domain block back to the operator, the amount to
+/// refund to a particular operator is determined by the total storage fee collected from the domain user
+/// and the percentage of bundle storage that the operator have submitted for the domain block.
+#[allow(dead_code)]
+pub fn refund_storage_fee<T: Config>(
+    total_storage_fee: BalanceOf<T>,
+    front_paid_storage: BTreeMap<OperatorId, u32>,
+) -> Result<(), Error> {
+    if total_storage_fee.is_zero() {
+        return Ok(());
+    }
+
+    let total_paid_storage = front_paid_storage.values().sum::<u32>();
+    let mut remaining_fee = total_storage_fee;
+    for (operator_id, paid_storage) in front_paid_storage {
+        // If the operator is deregistered and unlocked or slashed and finalized, the refund bundle storage
+        // fee will go to the treasury
+        if Operators::<T>::get(operator_id).is_none() || paid_storage.is_zero() {
+            continue;
+        }
+
+        let refund_amount = {
+            let share = Perbill::from_rational(paid_storage, total_paid_storage);
```

Comment:
Maybe we should introduce the same `SharePrice` struct here as well

Reply:
We don't have the same issue as the staking share here, we just need to compute a percentage of the total fee for each operator, if `share` is too confusing perhaps rename `share` to `percentage`?

Reply:
FYI @vedhavyas maybe [spec](https://www.notion.so/subspacelabs/Fees-Rewards-Specification-WIP-1b835c7684a940f188920802ca6791f2?pvs=4#6688692ad473489e81c348570300ded6) is a bit more clear. The bundle authors just get refunded proportionally to what they paid in a block. Totally independent of stake shares.

Reply:
Yes I understand that but I'm confused with the variable naming here as `share` instead of something like `paid_storage_percentage` or something that not overlapping with staking terminology